### DISCUSSION TITLE
The validation for type checking must be changed

### DISCUSSION THREAD

**Discussion by samislam** [Automattic/mongoose#12288] [Ideas]
I am having extra work to do on the front end because of missing built-in features which should be supported in Mongoose by default.

# Request [1]: Either allow **all** the fields to accept empty strings as a signal to reset the field to _empty_, or disallow it completely.

Take this example:

I have the following schema:
```js
const mySchema = new mongoose.Schema({
   name: String,
   room: {
      type: mongoose.Schema.ObjectId,
   },
   age: Number
})
```

And on the web front-end, I have the following inputs in an **Update User Form**  (used to update a user):
![Loading...](https://i.ibb.co/X4FDmSH/Screenshot-from-2022-08-17-02-26-59.png)
[click here to copy the code 🗒](https://pastebin.pl/view/07acbba3) 

I want to support the following features:
[*] All the fields are optional to be filled.
[*] If the field is empty, this means that the user wants to reset the field to empty, `(ex: name: "Jamaal" ---> name: "")`

Now, after testing that, the following happens:

### Form values:
![Loading...](https://i.ibb.co/0Gm1Mb2/Screenshot-from-2022-08-17-02-27-26.png)
### JSON:
```json
{
  "name": "Joe Nim",
  "age": "10",
  "room": "507f1f77bcf86cd799439011"
}
```
### Result:
as expected. the user document has been updated to hold the values in the JSON object.

### Form values:
![Loading...](https://i.ibb.co/TRHkLLj/Screenshot-from-2022-08-17-02-27-35.png)
### JSON:
```json
{
  "name": "",
  "age": "10",
  "room": "507f1f77bcf86cd799439011"
}
```
### Result:
The user document's name field has been updated to become an empty string in the database.

### Form values:
![Loading...](https://i.ibb.co/c6RMcTB/Screenshot-from-2022-08-17-02-27-42.png)
### JSON:
```json
{
  "name": "",
  "age": "",
  "room": "507f1f77bcf86cd799439011"
}
```
### Result:
The user document's name and age fields have been updated to become empty strings in the database.

### Form values:
![Loading...](https://i.ibb.co/0M8S8H8/Screenshot-from-2022-08-17-02-27-49.png)
### JSON:
```json
{
  "name": "",
  "age": "",
  "room": ""
}
```
### Result:
> **Cast Error:** Expected an Object Id but got a string instead.

Now, If I had to fix this problem and allow removing the **room** field, I would do that on the front-end as follows:

```js
function formSubmit(e){
  e.preventDefault()
  let formValues = new FormData(e.target)
  formValues = Object.fromEntries(formValues)
  if(formValues.room === "") formValues.room = null
}
```
* But why do I have to do that, I believe that should be handled somehow by Mongoose.

# Request [2]: Prevent allowing the Schema types String & Number (idk, and the others?) to accept `null` as a value.

Take the following example: 

I have the following schema:

```js
const mySchema = new mongoose.Schema({
blogPost: String
})
```
and I have the following express routes
```js
app.get('/posts/:postId', async (req, res, next)=>{
  const doc = Posts.findById(req.params.postId)
  res.status(200).json({
     data: doc
  })
})
app.patch('/posts/:postId', async (req, res, next)=>{
  const updatedDoc = Posts.findByIdAndUpdate(req.params.id, req.body, { runValidators: true, new: true })
  res.status(200).json({
     data: updatedDoc
  })
})
```

Let's pretend that I'm a pro admin who likes to write his blog posts using postman instead of the front-end webpage and I wrote the following in the body of the _PATCH_ request: 

![Loading...](https://i.ibb.co/kg3tMRM/Screenshot-from-2022-08-17-02-53-49.png)

And lets say that on the front-end, the code which is used to view the post looks as follows:
```html
<div class="view">   <!-- blog post content goes here --->   </div>
<script>
const view = document.querySelector('.view')
async function fetchPost(postId){
  const { data: post } = axios.get(`https://api.example.com/posts/${postId}`)
  const content = post.blogPost.capitalize() // 👈 here's the problem 💣
  view.innerText = content
}
// on page load
await fetchPost("507f191e810c19729de860ea")
</script>
```
In this example, the admin used Postman to perform an XSS attack without noticing, and now the front-end is having problems rendering the blog post, because it's of type **null** and `.capitalize()` is not a function on null values, it can only be used on strings.

This means that we need to find a better way to signal that _we want to reset this field to empty_ from the front end.

**Comment by vkarpov15** [comment]
So what is the question here? Is the issue just that passing an empty string `''` to a property that expects an ObjectId causes a CastError, or is there something else?

**Reply by samislam** [reply]
Yes. Just like that

**Reply by vkarpov15** [reply]
@samislam here's an example of how you can convert empty string to null for ObjectIds using [custom casting](https://mongoosejs.com/docs/tutorials/custom-casting.html). Does this help?

```javascript
'use strict';

const mongoose = require('mongoose');

// Special case for casting empty strings '': convert
// '' to `null` for ObjectIds
const castObjectId = mongoose.ObjectId.cast();
mongoose.ObjectId.cast(v => {
  if (v === '') {
    return null;
  }
  return castObjectId(v);
});

run().catch(err => {
  console.log(err);
  process.exit(-1);
});

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test');

  const schema = new mongoose.Schema({ testId: mongoose.ObjectId });
  const Test = mongoose.model('Test', schema);

  const { _id } = await Test.create({ testId: '' });

  const doc = await Test.findById(_id);
  console.log(doc);
}
```
